### PR TITLE
REF: simplify maybe_convert_objects

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also avoids some memory allocations which is nice.

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -1284,13 +1284,12 @@ cdef class Seen:
...
     @property
     def is_bool(self):
-        return not (self.datetime_ or self.numeric_ or self.timedelta_
-                    or self.nat_)
-
-    @property
-    def is_float_or_complex(self):
-        return not (self.bool_ or self.datetime_ or self.timedelta_
-                    or self.nat_)
+        # i.e. not (anything but bool)
```
Comment (mroeschke) [inline pandas/_libs/lib.pyx:1287]: Would be nice if this was `def anything_but_bool` as it's odd that `self.bool_ = False` would still return `True`. Maybe a followup

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Comment by lukemanley** [comment]
I think this may have caused a slowdown here: https://asv-runner.github.io/asv-collection/pandas/#reshape.Explode.time_explode

**Comment by jbrockmendel** [comment]
what do you think the culprit is?

**Comment by lukemanley** [comment]
actually, looks to be https://github.com/pandas-dev/pandas/pull/49566 - and there is some discussion there about adding a fix for the perf

**Pull request merged**